File: yupana-jdbc/src/main/java/org/yupana/jdbc/YupanaDatabaseMetaData.java
/*
 * Copyright 2019 Rusexpertiza LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.yupana.jdbc;

import java.sql.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;

import org.yupana.Proto;
import org.yupana.api.query.Result;
import org.yupana.api.query.SimpleResult;
import org.yupana.api.types.DataType;
import org.yupana.jdbc.build.BuildInfo;

public class YupanaDatabaseMetaData implements DatabaseMetaData {

    public YupanaDatabaseMetaData(YupanaConnection connection) {
        this.connection = connection;
    }

    private final YupanaConnection connection;

    private YupanaResultSet emptyResultSet() {
        return new YupanaResultSet(null, Result.empty());
    }

    @Override
    public boolean supportsMinimumSQLGrammar() throws SQLException {
        return false;
    }

    @Override
    public int getResultSetHoldability() throws SQLException {
        return ResultSet.CLOSE_CURSORS_AT_COMMIT;
    }

    @Override
    public int getMaxColumnNameLength() throws SQLException {
        return 0;
    }

    @Override
    public int getMaxColumnsInGroupBy() throws SQLException {
        return 0;
    }

    @Override
    public int getMaxColumnsInIndex() throws SQLException {
        return 0;
    }

    @Override
    public int getMaxColumnsInOrderBy() throws SQLException {
        return 0;
    }

    @Override
    public int getMaxColumnsInSelect() throws SQLException {
        return 0;
    }

    @Override
    public int getMaxColumnsInTable() throws SQLException {
        return 0;
    }

    @Override
    public boolean supportsSubqueriesInComparisons() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsSubqueriesInExists() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsSubqueriesInIns() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsSubqueriesInQuantifieds() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsCorrelatedSubqueries() throws SQLException {
        return false;
    }

    @Override
    public boolean nullPlusNonNullIsNull() throws SQLException {
        return true;
    }

    @Override
    public boolean supportsCatalogsInDataManipulation() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsCatalogsInProcedureCalls() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsCatalogsInTableDefinitions() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsCatalogsInIndexDefinitions() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsCatalogsInPrivilegeDefinitions() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsDataDefinitionAndDataManipulationTransactions() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsTableCorrelationNames() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsDifferentTableCorrelationNames() throws SQLException {
        return false;
    }

    @Override
    public int getDefaultTransactionIsolation() throws SQLException {
        return Connection.TRANSACTION_NONE;
    }

    @Override
    public boolean supportsTransactionIsolationLevel(int level) throws SQLException {
        return level == Connection.TRANSACTION_NONE;
    }

    @Override
    public boolean supportsOuterJoins() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsFullOuterJoins() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsLimitedOuterJoins() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsExpressionsInOrderBy() throws SQLException {
        return false;
    }

    @Override
    public boolean allProceduresAreCallable() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsResultSetConcurrency(int type, int concurrency) throws SQLException {
        return type == ResultSet.TYPE_FORWARD_ONLY && concurrency == ResultSet.CONCUR_READ_ONLY;
    }

    @Override
    public boolean nullsAreSortedHigh() throws SQLException {
        return false;
    }

    @Override
    public boolean nullsAreSortedLow() throws SQLException {
        return false;
    }

    @Override
    public boolean nullsAreSortedAtStart() throws SQLException {
        return false;
    }

    @Override
    public boolean nullsAreSortedAtEnd() throws SQLException {
        return false;
    }

    @Override
    public ResultSet getPrimaryKeys(String catalog, String schema, String table) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public ResultSet getImportedKeys(String catalog, String schema, String table) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public ResultSet getExportedKeys(String catalog, String schema, String table) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public boolean supportsPositionedDelete() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsPositionedUpdate() throws SQLException {
        return false;
    }

    @Override
    public boolean ownUpdatesAreVisible(int type) throws SQLException {
        return false;
    }

    @Override
    public boolean ownDeletesAreVisible(int type) throws SQLException {
        return false;
    }

    @Override
    public boolean ownInsertsAreVisible(int type) throws SQLException {
        return false;
    }

    @Override
    public boolean supportsResultSetHoldability(int holdability) throws SQLException {
        return holdability == ResultSet.CLOSE_CURSORS_AT_COMMIT;
    }

    @Override
    public int getMaxStatements() throws SQLException {
        return 0;
    }

    @Override
    public RowIdLifetime getRowIdLifetime() throws SQLException {
        return RowIdLifetime.ROWID_UNSUPPORTED;
    }

    @Override
    public String getSchemaTerm() throws SQLException {
        return "schema";
    }

    @Override
    public int getMaxCatalogNameLength() throws SQLException {
        return 0;
    }

    @Override
    public ResultSet getCrossReference(String parentCatalog, String parentSchema, String parentTable, String foreignCatalog, String foreignSchema, String foreignTable) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public String getCatalogTerm() throws SQLException {
        return "catalog";
    }

    @Override
    public ResultSet getAttributes(String catalog, String schemaPattern, String typeNamePattern, String attributeNamePattern) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public int getMaxStatementLength() throws SQLException {
        return 0;
    }

    @Override
    public boolean supportsBatchUpdates() throws SQLException {
        return false;
    }

    @Override
    public boolean allTablesAreSelectable() throws SQLException {
        return true;
    }

    @Override
    public int getMaxCharLiteralLength() throws SQLException {
        return 0;
    }

    @Override
    public boolean supportsMultipleOpenResults() throws SQLException {
        return true;
    }

    @Override
    public int getMaxRowSize() throws SQLException {
        return 0;
    }

    @Override
    public boolean supportsUnion() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsUnionAll() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsOpenCursorsAcrossCommit() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsOpenCursorsAcrossRollback() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsOpenStatementsAcrossCommit() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsOpenStatementsAcrossRollback() throws SQLException {
        return false;
    }

    @Override
    public String getSearchStringEscape() throws SQLException {
        return "\\";
    }

    @Override
    public int getMaxBinaryLiteralLength() throws SQLException {
        return 0;
    }

    @Override
    public boolean supportsAlterTableWithAddColumn() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsAlterTableWithDropColumn() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsResultSetType(int type) throws SQLException {
        return type == ResultSet.TYPE_FORWARD_ONLY;
    }

    @Override
    public boolean supportsSelectForUpdate() throws SQLException {
        return false;
    }

    @Override
    public ResultSet getUDTs(String catalog, String schemaPattern, String typeNamePattern, int[] types) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public String getSystemFunctions() throws SQLException {
        return "";
    }

    @Override
    public boolean supportsColumnAliasing() throws SQLException {
        return true;
    }

    @Override
    public boolean deletesAreDetected(int type) throws SQLException {
        return false;
    }

    @Override
    public boolean insertsAreDetected(int type) throws SQLException {
        return false;
    }

    @Override
    public boolean supportsMixedCaseIdentifiers() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsMixedCaseQuotedIdentifiers() throws SQLException {
        return false;
    }

    @Override
    public String getDatabaseProductName() throws SQLException {
        return "Yupana";
    }

    @Override
    public String getDatabaseProductVersion() throws SQLException {
        return connection.getServerVersion().map(Proto.Version::getVersion).orElse("Unknown");
    }

    @Override
    public int getDatabaseMajorVersion() throws SQLException {
        return connection.getServerVersion().map(Proto.Version::getMajor).orElse(0);
    }

    @Override
    public int getDatabaseMinorVersion() throws SQLException {
        return connection.getServerVersion().map(Proto.Version::getMinor).orElse(0);
    }

    @Override
    public String getSQLKeywords() throws SQLException {
        return "";
    }

    @Override
    public boolean dataDefinitionCausesTransactionCommit() throws SQLException {
        return false;
    }

    @Override
    public boolean dataDefinitionIgnoredInTransactions() throws SQLException {
        return false;
    }

    @Override
    public int getJDBCMajorVersion() throws SQLException {
        return 4;
    }

    @Override
    public int getJDBCMinorVersion() throws SQLException {
        return 1;
    }

    @Override
    public boolean isReadOnly() throws SQLException {
        return true;
    }

    @Override
    public ResultSet getProcedureColumns(String catalog, String schemaPattern, String procedureNamePattern, String columnNamePattern) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public ResultSet getCatalogs() throws SQLException {
        return emptyResultSet();
    }

    @Override
    public boolean locatorsUpdateCopy() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsANSI92EntryLevelSQL() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsANSI92IntermediateSQL() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsANSI92FullSQL() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsMultipleResultSets() throws SQLException {
        return false;
    }

    @Override
    public String getDriverName() throws SQLException {
        return "Yupana JDBC";
    }

    @Override
    public String getDriverVersion() throws SQLException {
        return BuildInfo.version();
    }

    @Override
    public int getDriverMajorVersion() {
        return BuildInfo.majorVersion();
    }

    @Override
    public int getDriverMinorVersion() {
        return BuildInfo.minorVersion();
    }

    @Override
    public int getMaxConnections() throws SQLException {
        return 0;
    }

    @Override
    public boolean othersUpdatesAreVisible(int type) throws SQLException {
        return false;
    }

    @Override
    public boolean othersDeletesAreVisible(int type) throws SQLException {
        return false;
    }

    @Override
    public boolean othersInsertsAreVisible(int type) throws SQLException {
        return false;
    }

    @Override
    public ResultSet getVersionColumns(String catalog, String schema, String table) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public boolean supportsNamedParameters() throws SQLException {
        return false;
    }

    @Override
    public boolean doesMaxRowSizeIncludeBlobs() throws SQLException {
        return false;
    }

    @Override
    public ResultSet getBestRowIdentifier(String catalog, String schema, String table, int scope, boolean nullable) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public ResultSet getProcedures(String catalog, String schemaPattern, String procedureNamePattern) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public int getMaxSchemaNameLength() throws SQLException {
        return 0;
    }

    @Override
    public boolean supportsIntegrityEnhancementFacility() throws SQLException {
        return false;
    }

    @Override
    public ResultSet getTablePrivileges(String catalog, String schemaPattern, String tableNamePattern) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public boolean supportsExtendedSQLGrammar() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsConvert() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsConvert(int fromType, int toType) throws SQLException {
        return false;
    }

    @Override
    public boolean autoCommitFailureClosesAllResultSets() throws SQLException {
        return false;
    }

    @Override
    public ResultSet getSuperTypes(String catalog, String schemaPattern, String typeNamePattern) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public boolean supportsMultipleTransactions() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsStatementPooling() throws SQLException {
        return false;
    }

    @Override
    public int getMaxCursorNameLength() throws SQLException {
        return 0;
    }

    @Override
    public String getUserName() throws SQLException {
        return "";
    }

    @Override
    public boolean supportsDataManipulationTransactionsOnly() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsLikeEscapeClause() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsGetGeneratedKeys() throws SQLException {
        return false;
    }

    @Override
    public ResultSet getIndexInfo(String catalog, String schema, String table, boolean unique, boolean approximate) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public ResultSet getSchemas() throws SQLException {
        return emptyResultSet();
    }

    @Override
    public ResultSet getSchemas(String catalog, String schemaPattern) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public String getCatalogSeparator() throws SQLException {
        return "";
    }

    @Override
    public String getExtraNameCharacters() throws SQLException {
        return "";
    }

    @Override
    public String getURL() throws SQLException {
        return connection.getUrl();
    }

    @Override
    public boolean supportsTransactions() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsStoredProcedures() throws SQLException {
        return false;
    }

    @Override
    public String getIdentifierQuoteString() throws SQLException {
        return "\"";
    }

    @Override
    public int getMaxTableNameLength() throws SQLException {
        return 0;
    }

    @Override
    public int getSQLStateType() throws SQLException {
        return DatabaseMetaData.sqlStateSQL;
    }

    @Override
    public ResultSet getColumnPrivileges(String catalog, String schema, String table, String columnNamePattern) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public ResultSet getTableTypes() throws SQLException {
        Collection<String> names = Collections.singleton("TABLE_TYPE");
        Collection<DataType> dataTypes = Collections.singleton(DataType.stringDt());
        Result types = new SimpleResult("TYPES", names, dataTypes, Collections.singleton("TABLE").iterator()))
        new YupanaResultSet(null, types);
    }

    @Override
    public boolean supportsNonNullableColumns() throws SQLException {
        return false;
    }

    @Override
    public int getMaxUserNameLength() throws SQLException {
        return 0;
    }

    @Override
    public boolean supportsSavepoints() throws SQLException {
        return false;
    }

    @Override
    public ResultSet getSuperTables(String catalog, String schemaPattern, String tableNamePattern) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public String getTimeDateFunctions() throws SQLException {
        return getFunctionsForType("TIMESTAMP");
    }

    @Override
    public String getNumericFunctions() throws SQLException {
        return getFunctionsForType("DECIMAL");
    }

    @Override
    public String getStringFunctions() throws SQLException {
        return getFunctionsForType("VARCHAR");
    }

    @Override
    public boolean generatedKeyAlwaysReturned() throws SQLException {
        return false;
    }

    @Override
    public boolean isCatalogAtStart() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsCoreSQLGrammar() throws SQLException {
        return false;
    }

    @Override
    public int getMaxProcedureNameLength() throws SQLException {
        return 0;
    }

    @Override
    public int getMaxTablesInSelect() throws SQLException {
        return 0;
    }

    @Override
    public ResultSet getColumns(String catalog, String schemaPattern, String tableNamePattern, String columnNamePattern) throws SQLException {
        String sql = "SHOW COLUMNS FROM " + tableNamePattern;
        Statement stmt = connection.createStatement();
        return stmt.executeQuery(sql);
    }

    @Override
    public Connection getConnection() throws SQLException {
        return connection;
    }

    @Override
    public ResultSet getTables(String catalog, String schemaPattern, String tableNamePattern, String[] types) throws SQLException {
        String sql = "SHOW TABLES";
        Statement stmt = connection.createStatement();
        return stmt.executeQuery(sql);
    }

    @Override
    public boolean supportsGroupBy() throws SQLException {
        return true;
    }

    @Override
    public boolean supportsGroupByUnrelated() throws SQLException {
        return true;
    }

    @Override
    public boolean supportsGroupByBeyondSelect() throws SQLException {
        return true;
    }

    @Override
    public boolean supportsOrderByUnrelated() throws SQLException {
        return false;
    }

    @Override
    public int getMaxIndexLength() throws SQLException {
        return 0;
    }

    @Override
    public String getProcedureTerm() throws SQLException {
        return "procedure";
    }

    @Override
    public ResultSet getFunctions(String catalog, String schemaPattern, String functionNamePattern) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public ResultSet getFunctionColumns(String catalog, String schemaPattern, String functionNamePattern, String columnNamePattern) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public ResultSet getClientInfoProperties() throws SQLException {
        return emptyResultSet();
    }

    @Override
    public boolean supportsStoredFunctionsUsingCallSyntax() throws SQLException {
        return false;
    }

    @Override
    public ResultSet getPseudoColumns(String catalog, String schemaPattern, String tableNamePattern, String columnNamePattern) throws SQLException {
        return emptyResultSet();
    }

    @Override
    public boolean usesLocalFiles() throws SQLException {
        return false;
    }

    @Override
    public boolean usesLocalFilePerTable() throws SQLException {
        return false;
    }

    @Override
    public boolean updatesAreDetected(int type) throws SQLException {
        return false;
    }

    @Override
    public ResultSet getTypeInfo() throws SQLException {
        return emptyResultSet();
    }

    @Override
    public boolean supportsSchemasInDataManipulation() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsSchemasInProcedureCalls() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsSchemasInTableDefinitions() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsSchemasInIndexDefinitions() throws SQLException {
        return false;
    }

    @Override
    public boolean supportsSchemasInPrivilegeDefinitions() throws SQLException {
        return false;
    }

    @Override
    public boolean storesUpperCaseIdentifiers() throws SQLException {
        return false;
    }

    @Override
    public boolean storesLowerCaseIdentifiers() throws SQLException {
        return true;
    }

    @Override
    public boolean storesMixedCaseIdentifiers() throws SQLException {
        return false;
    }

    @Override
    public boolean storesUpperCaseQuotedIdentifiers() throws SQLException {
        return false;
    }

    @Override
    public boolean storesLowerCaseQuotedIdentifiers() throws SQLException {
        return true;
    }

    @Override
    public boolean storesMixedCaseQuotedIdentifiers() throws SQLException {
        return false;
    }

    public <T> T unwrap(Class<T> aClass) throws SQLException {
        if (!aClass.isAssignableFrom(getClass())) {
            throw new SQLException(String.format("Cannot unwrap to %s", aClass.getName()));
        }

        return aClass.cast(this);
    }

    @Override
    public boolean isWrapperFor(Class<?> iface) throws SQLException {
        return iface.isAssignableFrom(getClass());
    }

    private String getFunctionsForType(String t) throws SQLException {
        String sql = "SHOW FUNCTIONS FOR" + t;
        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);

        ArrayList<String> functions = new ArrayList<>();
        while (rs.next()) {
            functions.add(rs.getString("NAME"));
        }

        return String.join(",", functions);
    }
}

File: yupana-jdbc/src/main/scala/org/yupana/jdbc/YupanaDatabaseMetaData.scala
/*
 * Copyright 2019 Rusexpertiza LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.yupana.jdbc

import java.sql.{ Connection, DatabaseMetaData, ResultSet, RowIdLifetime, SQLException }

import org.yupana.api.query.{ Result, SimpleResult }
import org.yupana.api.types.DataType
import org.yupana.jdbc.build.BuildInfo

class YupanaDatabaseMetaData(connection: YupanaConnection) extends DatabaseMetaData {

  private def emptyResultSet = new YupanaResultSet(null, Result.empty)

  override def supportsMinimumSQLGrammar(): Boolean = false

  override def getResultSetHoldability: Int = ResultSet.CLOSE_CURSORS_AT_COMMIT

  override def getMaxColumnsInGroupBy: Int = 0

  override def supportsSubqueriesInComparisons(): Boolean = false

  override def getMaxColumnsInSelect: Int = 0

  override def nullPlusNonNullIsNull() = true

  override def supportsCatalogsInDataManipulation() = false

  override def supportsDataDefinitionAndDataManipulationTransactions() = false

  override def supportsTableCorrelationNames() = false

  override def getDefaultTransactionIsolation: Int = Connection.TRANSACTION_NONE

  override def supportsFullOuterJoins(): Boolean = false

  override def supportsExpressionsInOrderBy() = false

  override def allProceduresAreCallable() = false

  override def supportsResultSetConcurrency(`type`: Int, concurrency: Int): Boolean =
    `type` == ResultSet.TYPE_FORWARD_ONLY && concurrency == ResultSet.CONCUR_READ_ONLY

  override def getMaxTablesInSelect = 1

  override def nullsAreSortedAtStart() = false

  override def getImportedKeys(catalog: String, schema: String, table: String): ResultSet = emptyResultSet

  override def supportsPositionedUpdate() = false

  override def ownDeletesAreVisible(`type`: Int) = false

  override def supportsResultSetHoldability(holdability: Int): Boolean =
    holdability == ResultSet.CLOSE_CURSORS_AT_COMMIT

  override def getMaxStatements = 0

  override def getRowIdLifetime = RowIdLifetime.ROWID_UNSUPPORTED

  override def getSchemaTerm = "schema"

  override def getMaxCatalogNameLength = 0

  override def getCrossReference(
      parentCatalog: String,
      parentSchema: String,
      parentTable: String,
      foreignCatalog: String,
      foreignSchema: String,
      foreignTable: String
  ): ResultSet = emptyResultSet

  override def getCatalogTerm = "catalog"

  override def getAttributes(
      catalog: String,
      schemaPattern: String,
      typeNamePattern: String,
      attributeNamePattern: String
  ): ResultSet = emptyResultSet

  override def getMaxStatementLength = 0

  override def supportsOuterJoins() = false

  override def supportsBatchUpdates() = false

  override def supportsLimitedOuterJoins() = false

  override def getMaxColumnsInTable = 0

  override def allTablesAreSelectable() = true

  override def getMaxCharLiteralLength = 0

  override def supportsMultipleOpenResults() = true

  override def getMaxRowSize = 0

  override def supportsUnion() = false

  override def supportsOpenCursorsAcrossCommit() = false

  override def ownUpdatesAreVisible(`type`: Int) = false

  override def getSearchStringEscape = "\\"

  override def getMaxBinaryLiteralLength = 0

  override def supportsAlterTableWithDropColumn = false

  override def supportsResultSetType(`type`: Int): Boolean = `type` == ResultSet.TYPE_FORWARD_ONLY

  override def supportsCatalogsInProcedureCalls() = false

  override def supportsSelectForUpdate() = false

  override def getUDTs(catalog: String, schemaPattern: String, typeNamePattern: String, types: Array[Int]): ResultSet =
    emptyResultSet

  override def supportsOpenStatementsAcrossRollback() = false

  override def getSystemFunctions = ""

  override def supportsColumnAliasing() = true

  override def insertsAreDetected(`type`: Int) = false

  override def supportsMixedCaseIdentifiers() = false

  override def getDatabaseProductVersion: String = connection.serverVersion.map(_.version).getOrElse("Unknown")

  override def getDatabaseMajorVersion: Int = connection.serverVersion.map(_.major).getOrElse(0)

  override def getDatabaseMinorVersion: Int = connection.serverVersion.map(_.minor).getOrElse(0)

  override def getSQLKeywords = ""

  override def dataDefinitionIgnoredInTransactions = false

  override def getJDBCMajorVersion = 4

  override def getJDBCMinorVersion = 1

  override def getMaxColumnNameLength = 0

  override def isReadOnly = true

  override def getProcedureColumns(
      catalog: String,
      schemaPattern: String,
      procedureNamePattern: String,
      columnNamePattern: String
  ): ResultSet = emptyResultSet

  override def getCatalogs: ResultSet = emptyResultSet

  override def locatorsUpdateCopy = false

  override def supportsANSI92FullSQL() = false

  override def supportsMultipleResultSets() = false

  override def storesUpperCaseIdentifiers() = false

  override def supportsSchemasInPrivilegeDefinitions() = false

  override def getDriverName = "Yupana JDBC"

  override def getDriverVersion: String = BuildInfo.version

  override def getDriverMajorVersion: Int = BuildInfo.majorVersion

  override def getDriverMinorVersion: Int = BuildInfo.minorVersion

  override def getMaxConnections = 0

  override def othersUpdatesAreVisible(`type`: Int) = false

  override def getVersionColumns(catalog: String, schema: String, table: String): ResultSet = emptyResultSet

  override def supportsNamedParameters() = false

  override def doesMaxRowSizeIncludeBlobs = false

  override def getBestRowIdentifier(
      catalog: String,
      schema: String,
      table: String,
      scope: Int,
      nullable: Boolean
  ): ResultSet = emptyResultSet

  override def getProcedures(catalog: String, schemaPattern: String, procedureNamePattern: String): ResultSet =
    emptyResultSet

  override def supportsSubqueriesInQuantifieds() = false

  override def getMaxSchemaNameLength = 0

  override def supportsIntegrityEnhancementFacility() = false

  override def getTablePrivileges(catalog: String, schemaPattern: String, tableNamePattern: String): ResultSet =
    emptyResultSet

  override def supportsExtendedSQLGrammar() = false

  override def supportsConvert() = false

  override def supportsConvert(fromType: Int, toType: Int) = false

  override def getFunctionColumns(
      catalog: String,
      schemaPattern: String,
      functionNamePattern: String,
      columnNamePattern: String
  ): ResultSet = emptyResultSet

  override def supportsPositionedDelete() = false

  override def autoCommitFailureClosesAllResultSets() = false

  override def getMaxColumnsInOrderBy = 0

  override def supportsANSI92IntermediateSQL() = false

  override def getSuperTypes(catalog: String, schemaPattern: String, typeNamePattern: String): ResultSet =
    emptyResultSet

  override def supportsMultipleTransactions() = false

  override def supportsCatalogsInTableDefinitions() = false

  override def supportsOpenCursorsAcrossRollback() = false

  override def supportsStatementPooling() = false

  override def usesLocalFiles() = false

  override def storesMixedCaseQuotedIdentifiers() = false

  override def othersInsertsAreVisible(`type`: Int) = false

  override def supportsSchemasInProcedureCalls() = false

  override def getMaxCursorNameLength = 0

  override def getUserName = ""

  override def supportsTransactionIsolationLevel(level: Int): Boolean = level == Connection.TRANSACTION_NONE

  override def deletesAreDetected(`type`: Int) = false

  override def supportsDataManipulationTransactionsOnly() = false

  override def supportsLikeEscapeClause() = false

  override def supportsSchemasInDataManipulation() = false

  override def supportsGetGeneratedKeys() = false

  override def supportsGroupBy() = true

  override def getIndexInfo(
      catalog: String,
      schema: String,
      table: String,
      unique: Boolean,
      approximate: Boolean
  ): ResultSet = emptyResultSet

  override def getSchemas: ResultSet = emptyResultSet

  override def getSchemas(catalog: String, schemaPattern: String): ResultSet = emptyResultSet

  override def supportsSchemasInIndexDefinitions() = false

  override def getCatalogSeparator = ""

  override def getExtraNameCharacters = ""

  override def getURL: String = connection.url

  override def supportsDifferentTableCorrelationNames() = false

  override def supportsTransactions() = false

  override def storesLowerCaseQuotedIdentifiers() = true

  override def supportsANSI92EntryLevelSQL() = false

  override def supportsStoredProcedures() = false

  override def supportsCatalogsInPrivilegeDefinitions() = false

  override def getIdentifierQuoteString = "\""

  override def getMaxTableNameLength = 0

  override def getSQLStateType: Int = DatabaseMetaData.sqlStateSQL

  override def getColumnPrivileges(
      catalog: String,
      schema: String,
      table: String,
      columnNamePattern: String
  ): ResultSet = emptyResultSet

  override def getMaxColumnsInIndex = 0

  override def getTableTypes: ResultSet = {
    val names = List("TABLE_TYPE")
    val dataTypes = List(DataType[String])
    val types = SimpleResult("TYPES", names, dataTypes, Iterator(Array[Any]("TABLE"), Array[Any]("ROLLUP")))
    new YupanaResultSet(null, types)
  }

  override def nullsAreSortedHigh() = false

  override def supportsNonNullableColumns() = false

  override def getMaxUserNameLength = 0

  override def supportsSubqueriesInExists() = false

  override def ownInsertsAreVisible(`type`: Int): Boolean = false

  override def supportsOpenStatementsAcrossCommit() = false

  override def supportsSavepoints() = false

  override def getSuperTables(catalog: String, schemaPattern: String, tableNamePattern: String): ResultSet =
    emptyResultSet

  override def dataDefinitionCausesTransactionCommit = false

  override def nullsAreSortedAtEnd() = false

  override def getTimeDateFunctions: String = getFunctionsForType("TIMESTAMP")

  override def getNumericFunctions: String = getFunctionsForType("DECIMAL")

  override def getStringFunctions: String = getFunctionsForType("VARCHAR")

  override def generatedKeyAlwaysReturned = false

  override def supportsUnionAll() = false

  override def supportsAlterTableWithAddColumn() = false

  override def isCatalogAtStart = false

  override def othersDeletesAreVisible(`type`: Int) = false

  override def supportsCoreSQLGrammar() = false

  override def getMaxProcedureNameLength = 0

  override def getColumns(
      catalog: String,
      schemaPattern: String,
      tableNamePattern: String,
      columnNamePattern: String
  ): ResultSet = {
    val sql = s"SHOW COLUMNS FROM $tableNamePattern"
    val stmt = connection.createStatement()
    stmt.executeQuery(sql)
  }

  override def getConnection: Connection = connection

  override def getDatabaseProductName = "Yupana"

  override def supportsGroupByUnrelated() = true

  override def nullsAreSortedLow() = false

  override def getTables(
      catalog: String,
      schemaPattern: String,
      tableNamePattern: String,
      types: Array[String]
  ): ResultSet = {
    val sql = "SHOW TABLES"
    val stmt = connection.createStatement()
    stmt.executeQuery(sql)
  }

  override def supportsCorrelatedSubqueries() = false

  override def supportsMixedCaseQuotedIdentifiers() = false

  override def supportsGroupByBeyondSelect() = true

  override def supportsCatalogsInIndexDefinitions() = false

  override def supportsOrderByUnrelated() = false

  override def getMaxIndexLength = 0

  override def getProcedureTerm = "procedure"

  override def getFunctions(catalog: String, schemaPattern: String, functionNamePattern: String): ResultSet =
    emptyResultSet

  override def getClientInfoProperties: ResultSet = emptyResultSet

  override def supportsStoredFunctionsUsingCallSyntax() = false

  override def getPseudoColumns(
      catalog: String,
      schemaPattern: String,
      tableNamePattern: String,
      columnNamePattern: String
  ): ResultSet = emptyResultSet

  override def usesLocalFilePerTable() = false

  override def storesLowerCaseIdentifiers() = true

  override def supportsSubqueriesInIns() = false

  override def updatesAreDetected(`type`: Int) = false

  // TODO: Implement me
  override def getTypeInfo: ResultSet = emptyResultSet

  override def getExportedKeys(catalog: String, schema: String, table: String): ResultSet = emptyResultSet

  override def getPrimaryKeys(catalog: String, schema: String, table: String): ResultSet = emptyResultSet

  override def supportsSchemasInTableDefinitions() = false

  override def storesUpperCaseQuotedIdentifiers() = false

  override def storesMixedCaseIdentifiers() = false

  override def unwrap[T](iface: Class[T]): T = {
    if (!iface.isAssignableFrom(getClass)) {
      throw new SQLException(s"Cannot unwrap to ${iface.getName}")
    }

    iface.cast(this)
  }

  override def isWrapperFor(iface: Class[_]): Boolean = iface.isAssignableFrom(getClass)

  private def getFunctionsForType(t: String): String = {
    val sql = s"SHOW FUNCTIONS FOR $t"
    val stmt = connection.createStatement()
    val rs = stmt.executeQuery(sql)
    val fs = Iterator.continually(rs).takeWhile(_.next()).map(_.getString("NAME")).toSeq
    fs mkString ","
  }
}
